File: git/yaml_validation.rb
#!/usr/bin/env ruby
#
#
#
require 'yaml'
require 'tmpdir'

class YamlValidate
  def self.yaml_file?(filename)
    return filename.end_with?('.yaml') || filename.end_with?('.yml')
  end

  def self.short_yaml_ext?(filename)
    return filename.end_with?(".yml")
  end

  def self.valid_yaml?(filename)
    YAML::load_file(filename)

    return true
  end
end

class GitCommit
  attr_accessor :oldrev, :newrev, :refname, :tmp
  def initialize(oldrev, newrev, refname)
    @oldrev = oldrev
    @newrev = newrev
    @refname = refname
    @tmp = Dir.mktmpdir(@newrev)
  end

  def get_file_changes()
    files = %x[/usr/bin/git diff --name-only #{@oldrev} #{@newrev} --diff-filter=ACM].split("\n")

    # if files is empty we will get a full checkout.  This happens on
    # a git rm file.  If there are no changes then we need to skip the archive
    return [] if files.empty?

    # We only want to take the files that changed.  Archive will do that when passed
    # the filenames.  It will export these to a tmp dir
    system("/usr/bin/git archive #{@newrev} #{files.join(" ")} | tar x -C #{@tmp}")
    return Dir.glob("#{@tmp}/**/*").delete_if { |file| File.directory?(file) }
  end
end

if __FILE__ == $0
  while data = STDIN.gets
    oldrev, newrev, refname = data.split
    gc = GitCommit.new(oldrev, newrev, refname)

    results = []
    gc.get_file_changes().each do |file|
      begin
        puts "++++++ Received:  #{file}"

        #raise "Yaml file extensions must be .yaml not .yml" if YamlValidate.short_yaml_ext? file

        # skip readme, other files, etc
        next unless YamlValidate.yaml_file?(file)

        results << YamlValidate.valid_yaml?(file)
      rescue Exception => ex
        puts "\n#{ex.message}\n\n"
        results << false
      end
    end

    #puts "RESULTS\n#{results.inspect}\n"
    exit 1 if results.include?(false)
  end
end

File: git/yaml_validation.py
#!/usr/bin/env python
#
#  python yaml validator for a git commit
#
'''
python yaml validator for a git commit
'''
import shutil
import sys
import os
import tempfile
import subprocess
import yaml

def get_changes(oldrev, newrev, tempdir):
    '''Get a list of git changes from oldrev to newrev'''
    proc = subprocess.Popen(['/usr/bin/git', 'diff', '--name-only', oldrev,
                             newrev, '--diff-filter=ACM'], stdout=subprocess.PIPE)
    stdout, _ = proc.communicate()
    files = stdout.split('\n')

    # No file changes
    if not files:
        return []

    cmd = '/usr/bin/git archive %s %s | /bin/tar x -C %s' % (newrev, " ".join(files), tempdir)
    proc = subprocess.Popen(cmd, shell=True)
    _, _ = proc.communicate()

    rfiles = []
    for dirpath, _, fnames in os.walk(tempdir):
        for fname in fnames:
            rfiles.append(os.path.join(dirpath, fname))

    return rfiles

def main():
    '''
    Perform yaml validation
    '''
    results = []
    try:
        tmpdir = tempfile.mkdtemp(prefix='jenkins-git-')
        old, new, _ = sys.argv[1:]

        for file_mod in get_changes(old, new, tmpdir):

            print "+++++++ Received: %s" % file_mod

            # if the file extensions is not yml or yaml, move along.
            if not file_mod.endswith('.yml') and not file_mod.endswith('.yaml'):
                continue

            # We use symlinks in our repositories, ignore them.
            if os.path.islink(file_mod):
                continue

            try:
                yaml.load(open(file_mod))
                results.append(True)

            except yaml.scanner.ScannerError as yerr:
                print yerr
                results.append(False)
    finally:
        shutil.rmtree(tmpdir)

    if not all(results):
        sys.exit(1)

if __name__ == "__main__":
    main()

